### PR TITLE
feat: _winapi.GetModuleFileName

### LINKED ISSUE #4001: _winapi.GetModuleFileName
## Feature

_winapi.GetModuleFileName

## Python Source code
https://github.com/python/cpython/blob/v3.10.5/Modules/_winapi.c#L1292-L1324

### PR DESCRIPTION
**Pull request by jiang1997**
close #4001
close #3925 by including its all commits

### CODE REVIEW

File: stdlib/src/lib.rs [hunk 1]
```diff
@@ -133,6 +135,11 @@ pub fn get_module_inits() -> impl Iterator<Item = (Cow<'static, str>, StdlibInit
         {
             "_bz2" => bz2::make_module,
         }
+        // Windoes-only
+        #[cfg(windows)]
+        {
+            "win32api" => win32api::make_module,
+        }
```
Comment (youknowone) [inline stdlib/src/lib.rs:142]: `_winapi.GetModuleFileName` is part of `_winapi` module. We don't need to add our original module for the purpose.

**Comment by fanninpm** [comment]
You might be looking for "feat". ["feta" is a kind of cheese](https://en.wikipedia.org/wiki/Feta).

**Comment by youknowone** [comment]
> You might be looking for "feat". ["feta" is a kind of cheese](https://en.wikipedia.org/wiki/Feta).

@fanninpm you have permission to edit any title or text

File: vm/src/stdlib/winapi.rs [hunk 2]
```diff
@@ -402,4 +408,58 @@ mod _winapi {
...
+        fn from_wide_null(wide: &[u16]) -> Self;
+    }
+    impl FromWide for OsString {
+        fn from_wide_null(wide: &[u16]) -> OsString {
+            let len = wide.iter().take_while(|&&c| c != 0).count();
+            OsString::from_wide(&wide[..len])
+        }
+    }
+
+    const MAX_PATH: usize = 260;
```
Comment (jiang1997) [inline vm/src/stdlib/winapi.rs:436]: @youknowone Do you have any idea to set MAX_PATH?

Comment (youknowone) [reply vm/src/stdlib/winapi.rs:436]: I am not an expert of windows API, but reusing it from windows-rs?

https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/Foundation/constant.MAX_PATH.html

I found another article about it: https://docs.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry

Comment (jiang1997) [reply vm/src/stdlib/winapi.rs:436]: https://github.com/python/cpython/blob/v3.10.5/Modules/_winapi.c#L1309-L1324
I found that in the implementation of cpython, it uses ``WCHAR[MAX_PATH]``(``WCHAR = ctypes.c_wchar``
) and ``ctypes.c_wchar`` has a fixed size of 2 bytes https://github.com/nvaccess/nvda/issues/8981.
So it's fine to reuse ``MAX_PATH`` of windows-rs.

**Review by youknowone — CHANGES_REQUESTED** [review]
In general, looks very good. I left a few comments.
I will try to test this one with the http update because we don't have test for this function.

File: vm/src/stdlib/winapi.rs [hunk 3]
```diff
@@ -402,4 +406,58 @@ mod _winapi {
...
+
+    #[pyfunction]
+    fn LoadLibrary(path: PyStrRef, vm: &VirtualMachine) -> PyResult<isize> {
+        let path = path.as_str().to_wide_null();
+        let handle = unsafe { LoadLibraryW(PCWSTR::from_raw(path.as_ptr())).unwrap() };
+        if handle.is_invalid() {
+            return Err(vm.new_runtime_error("LoadLibrary failed".to_owned()));
+        }
+        Ok(handle.0)
+    }
```
[the remark this answers is not in the document]
Comment (jiang1997) [reply vm/src/stdlib/winapi.rs:448]: Sorry, I forget to mention that it's for my personal test only.
Before getting a module's file name, I must load someone first.

File: vm/src/stdlib/winapi.rs [hunk 4]
```diff
@@ -402,4 +406,58 @@ mod _winapi {
         })
         .map(drop)
     }
+
+    use std::ffi::{OsStr, OsString};
+    use std::os::windows::prelude::*;
```
[the remark this answers is not in the document]
Comment (jiang1997) [reply vm/src/stdlib/winapi.rs:412]: Got it.

**Review by youknowone — APPROVED** [review]
I merged it with the HTTP tests to check if it runs as expected.
Thank you for contributing!

**Pull request merged**